File: nextrace/src/query_parser.cpp
#include "query_parser.hpp"
#include <sstream>
#include <algorithm>
#include <cctype>
#include <regex>
#include <iostream>

static std::vector<std::string> shlex_split(const std::string& s, std::optional<std::string>& error) {
    std::vector<std::string> tokens;
    bool in_quotes = false;
    std::string current_token;
    
    for (size_t i = 0; i < s.length(); ++i) {
        char c = s[i];
        if (c == '"' || c == '\'') {
            if (in_quotes) {
                in_quotes = false;
            } else {
                in_quotes = true;
            }
        } else if (std::isspace(c) && !in_quotes) {
            if (!current_token.empty()) {
                tokens.push_back(current_token);
                current_token.clear();
            }
        } else {
            current_token += c;
        }
    }
    
    if (in_quotes) {
        error = "No closing quotation";
    }
    if (!current_token.empty()) {
        tokens.push_back(current_token);
    }
    
    return tokens;
}

ParsedQuery parse_query(const std::string& query_str) {
    ParsedQuery res;
    res.raw_query = query_str;
    res.op = Operator::NONE;

    if (query_str.empty() || std::all_of(query_str.begin(), query_str.end(), [](unsigned char c){ return std::isspace(c); })) {
        res.error = "Query string cannot be empty.";
        return res;
    }

    std::optional<std::string> split_error;
    std::vector<std::string> raw_parts = shlex_split(query_str, split_error);
    if (split_error) {
        res.error = "Malformed query: " + *split_error;
        return res;
    }

    if (raw_parts.empty()) {
        res.error = "Query string cannot be empty.";
        return res;
    }

    std::regex token_pattern(R"(\w+)");
    bool expected_operator = false;
    bool has_operator = false;

    for (size_t i = 0; i < raw_parts.size(); ++i) {
        std::string part = raw_parts[i];
        std::string part_upper = part;
        std::transform(part_upper.begin(), part_upper.end(), part_upper.begin(), ::toupper);

        if (part_upper == "AND" || part_upper == "OR") {
            if (!expected_operator) {
                if (i == 0) {
                    res.error = "Malformed query: invalid operator placement '" + part + "'.";
                    return res;
                } else {
                    res.error = "Malformed query: consecutive operators.";
                    return res;
                }
            }

            Operator current_op = (part_upper == "AND") ? Operator::AND : Operator::OR;
            if (!has_operator) {
                res.op = current_op;
                has_operator = true;
            } else if (res.op != current_op) {
                res.error = "Mixed AND/OR queries are not supported in this phase.";
                return res;
            }

            expected_operator = false;
        } else {
            if (expected_operator) {
                res.error = "Malformed query: missing operator between terms.";
                return res;
            }

            std::string part_lower = part;
            std::transform(part_lower.begin(), part_lower.end(), part_lower.begin(), ::tolower);

            auto words_begin = std::sregex_iterator(part_lower.begin(), part_lower.end(), token_pattern);
            auto words_end = std::sregex_iterator();

            for (std::sregex_iterator k = words_begin; k != words_end; ++k) {
                std::string pt = k->str();
                if (pt.length() >= 2 || std::all_of(pt.begin(), pt.end(), ::isdigit)) {
                    res.terms.push_back(pt);
                }
            }

            expected_operator = true;
        }
    }

    if (!expected_operator) {
        res.error = "Malformed query: trailing operator.";
        return res;
    }

    if (res.terms.empty()) {
        res.error = "Query string contains no valid terms.";
        return res;
    }

    if (res.terms.size() > 1 && !has_operator) {
        res.op = Operator::AND;
    }

    // Deduplicate terms while preserving order
    std::vector<std::string> unique_terms;
    for (const auto& t : res.terms) {
        if (std::find(unique_terms.begin(), unique_terms.end(), t) == unique_terms.end()) {
            unique_terms.push_back(t);
        }
    }
    res.terms = unique_terms;

    return res;
}

File: nextrace/query_parser.py
import shlex
import re

TOKEN_PATTERN = re.compile(r"\w+")

def parse_query(query_str: str) -> dict:
    """
    Parses a query string into a structured dictionary.
    
    Syntax supported:
    - Single term: `error`
    - Multi-term AND: `error AND timeout`
    - Multi-term OR: `error OR warning`
    
    Simplifications:
    - Quoted phrases (e.g., `"connection timeout"`) are tokenized into their constituent words. 
      If the query lacks an overarching operator, these words are treated as an implicit AND.
      However, if the overarching query uses OR (e.g., `error OR "connection timeout"`), 
      the constituent words are treated as part of the overarching OR query (i.e., `error OR connection OR timeout`),
      because mixed AND/OR operations are not supported in this phase.
    """
    if not query_str or not query_str.strip():
        raise ValueError("Query string cannot be empty.")
    
    try:
        raw_parts = shlex.split(query_str)
    except ValueError as e:
        raise ValueError(f"Malformed query: {e}")
        
    if not raw_parts:
        raise ValueError("Query string cannot be empty.")
        
    operator = None
    terms = []
    
    expected_operator = False
    
    for i, part in enumerate(raw_parts):
        part_upper = part.upper()
        if part_upper in ("AND", "OR"):
            if not expected_operator:
                if i == 0:
                    raise ValueError(f"Malformed query: invalid operator placement '{part}'.")
                else:
                    raise ValueError("Malformed query: consecutive operators.")
            
            if operator is None:
                operator = part_upper
            elif operator != part_upper:
                raise ValueError("Mixed AND/OR queries are not supported in this phase.")
            
            expected_operator = False
        else:
            if expected_operator:
                raise ValueError("Malformed query: missing operator between terms.")
            
            # Normalize and extract terms identical to tokenizer.py
            part_terms = TOKEN_PATTERN.findall(part.lower())
            for pt in part_terms:
                if len(pt) >= 2 or pt.isnumeric():
                    terms.append(pt)
            
            expected_operator = True
            
    if not expected_operator:
        raise ValueError("Malformed query: trailing operator.")
        
    if not terms:
        raise ValueError("Query string contains no valid terms.")
        
    # Default to AND if there are multiple terms but no operator was parsed
    # (e.g. a single quoted phrase with multiple words: `"connection timeout"`)
    if len(terms) > 1 and operator is None:
        operator = "AND"
        
    # Deduplicate terms while preserving order
    unique_terms = []
    for t in terms:
        if t not in unique_terms:
            unique_terms.append(t)
            
    return {
        "terms": unique_terms,
        "operator": operator,
        "raw_query": query_str
    }
